content/blog/similar-document-search-using-apache-spark-with-tf-idf.md

```diff
@@ -0,0 +1,167 @@
+---
+title: Similar Document Search using Apache Spark with TF-IDF
+date: 2021-06-15T06:09:32.919Z
+author: Prasad Singathi, Maikel Pereira
+authorimage: /img/maikel-pereira.jpg
+tags:
+  - MapR
+  - HPE Ezmeral Data Fabric
+  - Machine learning
+---
+**Editor’s Note:** MapR products and solutions sold prior to the acquisition of such assets by Hewlett Packard Enterprise Company in 2019 may have older product names and model numbers that differ from current solutions. For information about current offerings, which are now part of HPE Ezmeral Data Fabric, please visit <https://www.hpe.com/us/en/software/data-fabric.html>
+
+## Original Post Information:
+
+```
+"authorDisplayName": ["Prasad Singathi","Maikel Pereira"],
+"publish": "2019-06-18T07:00:00.000Z",
+"category": ["machine-learning"],
+```
+
+- - -
+
+## Background
+
+As a professional services group, we were tasked with providing a solution to automatically find messages in the archives that are similar to new messages and send them to the person asking the question. 
+
+To accomplish that goal, we decided to apply machine learning to the process, so that there is an automated program able to find similarities between the current message and the historical data. The algorithm used was *term frequency&mdash;inverse document frequency* (TF-IDF). TF-IDF is used in a variety of applications. Typical use cases include document search, document tagging, and finding similar documents.  
+
+## Problem Description
+
+The desired solution was built using two Apache Spark applications running in a MapR cluster: one of them uses the historical data to update data features and train the model on a regular basis, and the second one analyzes every new message and finds five similar ones.
+
+## Application 1 - Creates Features and Trains Model
+
+This application was developed using Spark and Scala, and it can run on a schedule, depending on the needs. **Here is what it does, step by step:**
+
+1. Loads all messages from MapR Database. For the sake of brevity, we omit preprocessing steps like tokenization, stop words removal, punctuation removal, and other types of cleanup.
+
+```
```

Comment:
For all the scala code block snippets throughout the post, please specify the language syntax ```scala